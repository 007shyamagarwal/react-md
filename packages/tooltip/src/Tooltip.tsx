import * as React from "react";
import * as PropTypes from "prop-types";
import cn from "classnames";

type TooltipMouseListeners = "mouseenter" | "mouseleave";

export interface ITooltipProps {
  id: string;
  visible: boolean;
  delay?: number;
  onShow: () => void;
  onHide: () => void;
  style?: React.CSSProperties;
  className?: string;
  children?: React.ReactNode;
}

export interface ITooltipDefaultProps {
  delay: number;
}

export type TooltipWithDefaultProps = ITooltipProps & ITooltipDefaultProps;

export interface ITooltipState {
  animatingIn: boolean;
  animatingOut: boolean;
}

export default class Tooltip extends React.Component<ITooltipProps, ITooltipState> {
  public static propTypes = {
    id: PropTypes.string.isRequired,
    style: PropTypes.object,
    className: PropTypes.string,
    children: PropTypes.node.isRequired,
    delay: PropTypes.number,
  };

  public static defaultProps: ITooltipDefaultProps = {
    delay: 0,
  };

  private container: Element | null;
  private timeout: number | null;

  constructor(props: TooltipWithDefaultProps) {
    super(props);

    this.container = null;
    this.timeout = null;
    this.state = {
      animatingIn: false,
      animatingOut: false,
    };
  }

  public componentDidMount() {
    this.init();
  }

  public componentWillUnmount() {
    this.deinit();
  }

  public render() {
    const { animatingIn, animatingOut } = this.state;
    const {
      id,
      style,
      className,
      visible,
      children,
    } = this.props;

    return (
      <span
        id={id}
        role="tooltip"
        style={style}
        className={cn("md-tooltip", {
          "md-tooltip--active": visible,
          "md-tooltip--enter": animatingIn,
          "md-tooltip--leave": animatingOut,
          "md-tooltip--visible": visible || animatingIn || animatingOut,
        }, className)}
        onTransitionEnd={this.handleTransitionEnd}
      >
        {children}
      </span>
    );
  }

  private init = () => {
    const { id } = this.props;
    const container = document.querySelector(`[aria-describedby="${id}"]`);
    if (!container) {
      throw new Error((
        "Unable to find a valid container for your tooltip."
      ));
    }

    container.addEventListener("mouseenter", this.handleMouseEnter);
    container.addEventListener("mouseleave", this.handleMouseLeave);
    // I can't get the keyboard handlers working correctly..
    container.addEventListener("keydown", this.handleKeyDown as any);
    container.addEventListener("keyup", this.handleKeyUp as any);
    this.container = container;
  }

  private deinit = () => {
    if (!this.container) {
      return;
    }

    this.container.removeEventListener("mouseenter", this.handleMouseEnter);
    this.container.removeEventListener("mouseleave", this.handleMouseLeave);
    this.container.removeEventListener("keydown", this.handleKeyDown as any);
    this.container.removeEventListener("keyup", this.handleKeyUp as any);
  }

  private clearTimeout = () => {
    if (this.timeout) {
      window.clearTimeout(this.timeout);
      this.timeout = null;
    }
  }

  private show = () => {
    if (this.timeout) {
      return;
    }

    const { delay, onShow } = this.props as TooltipWithDefaultProps;
    let afterUpdate;
    if (delay <= 0) {
      afterUpdate = () => onShow();
    } else {
      this.timeout = window.setTimeout(() => {
        onShow();
        this.timeout = null;
      }, delay);
    }

    this.setState({ animatingIn: true, animatingOut: false }, afterUpdate);
  }

  private hide = () => {
    const { onHide, visible } = this.props;
    this.clearTimeout();
    if (this.props.visible) {
      this.setState({ animatingOut: true, animatingIn: false }, () => onHide());
    }
  }

  private handleMouseEnter = () => {
    this.show();
  }

  private handleMouseLeave = () => {
    this.hide();
  }

  private handleKeyDown = (e: KeyboardEvent) => {
    if (e.key === "Escape") {
      this.hide();
    }
  }

  private handleKeyUp = (e: KeyboardEvent) => {
    if (e.key === "TabKey") {
      this.show();
    }
  }

  private handleTransitionEnd = (e: React.TransitionEvent<HTMLSpanElement>) => {
    const { visible } = this.props;
    const { animatingIn, animatingOut } = this.state;
    if (visible === animatingIn && animatingOut) {
      this.setState({ animatingOut: false });
    } else if (visible === animatingOut && animatingIn) {
      this.setState({ animatingIn: false });
    }
  }
}
