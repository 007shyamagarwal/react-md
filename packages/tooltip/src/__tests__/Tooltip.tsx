import * as React from "react";
import { shallow } from "enzyme";

import { default as Tooltip } from "../Tooltip";

describe("<Tooltip />", () => {
  it("should default to setting the tooltip position to bottom", () => {
    const tooltip = shallow(<Tooltip id="tooltip-1" visible={true}>This is a tooltip</Tooltip>);
    expect(tooltip.hasClass("rmd-tooltip--top")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--bottom")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);
  });

  it("should correctly apply the position suffix", () => {
    const tooltip = shallow(<Tooltip id="tooltip-1" position="top" visible={true}>This is a tooltip</Tooltip>);
    expect(tooltip.hasClass("rmd-tooltip--top")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--bottom")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    tooltip.setProps({ position: "right" });
    expect(tooltip.hasClass("rmd-tooltip--top")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--bottom")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    tooltip.setProps({ position: "bottom" });
    expect(tooltip.hasClass("rmd-tooltip--top")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--bottom")).toBe(true);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(false);

    tooltip.setProps({ position: "left" });
    expect(tooltip.hasClass("rmd-tooltip--top")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--right")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--bottom")).toBe(false);
    expect(tooltip.hasClass("rmd-tooltip--left")).toBe(true);
  });

  it("should render correctly", () => {
    const tooltip = shallow(<Tooltip id="tooltip-1" visible={false}>This is some content</Tooltip>);
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.setProps({ visible: true });
    expect(tooltip.render()).toMatchSnapshot();

    tooltip.setProps({ dense: true });
    expect(tooltip.render()).toMatchSnapshot();
  })
})
