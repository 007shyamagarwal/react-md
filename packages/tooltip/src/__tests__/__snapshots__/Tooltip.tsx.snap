// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render correctly 1`] = `
<span
  class="rmd-tooltip rmd-tooltip--bottom"
  id="tooltip-1"
  role="tooltip"
>
  This is some content
</span>
`;

exports[`<Tooltip /> should render correctly 2`] = `
<span
  class="rmd-tooltip rmd-tooltip--visible rmd-tooltip--bottom"
  id="tooltip-1"
  role="tooltip"
>
  This is some content
</span>
`;

exports[`<Tooltip /> should render correctly 3`] = `
<span
  class="rmd-tooltip rmd-tooltip--visible rmd-tooltip--dense rmd-tooltip--bottom"
  id="tooltip-1"
  role="tooltip"
>
  This is some content
</span>
`;
